core/src/main/scala/kafka/log/LogLoader.scala

```diff
@@ -418,10 +423,12 @@ object LogLoader extends Logging {
           try {
             recoverSegment(segment, params)
           } catch {
-            case _: InvalidOffsetException =>
+            case ooe: LogSegmentOffsetOverflowException =>
```

Comment:
Your concern is valid. I've changed the logic so that the log truncation will only happen when it's InvalidOffsetException or when the recovery flog is turned on.